mathoxide-lib/src/utils.rs

```diff
@@ -0,0 +1,77 @@
+pub struct ByRowIterator {
+    shape: Vec<usize>,
+    current: usize,
+    numel: usize,
+}
+
+impl Iterator for ByRowIterator {
+    type Item = Vec<usize>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.current >= self.numel {
+            None
+        } else {
+            let v = inverse_translate(self.current, &self.shape);
+            self.current += 1;
+            Some(v)
+        }
+    }
```

Comment:
```suggestion
fn next(&mut self) -> Option<Self::Item> {
    let ndim = self.shape.len();
    let mut carry: usize = 1;
    let mut pointer: usize = 0;
    while carry == 1 && pointer < ndim {
        self.current[ndim - pointer - 1] += carry;
        carry = self.current[ndim - pointer - 1] % self.shape[ndim - pointer - 1];
        pointer += 1;
    }
    if carry == 1 {
        None
    }
    else {
        Some(self.current.clone())
    }
}
```
This is another way to do it. `self.current` here is also a Vec<usize> and `numel` is not required. Keep whichever you think is easier to read/more idiomatic  for rust. Disclaimer: the code above is not tested